ch_pipeline/analysis/flagging.py

```diff
@@ -1702,6 +1702,348 @@ def process(self, ss):
+    max_frac_freq: float
+        Mask any frequency that was transmitted by an FPGA motherboard slot
+        with more than this fraction of frequencies masked by the above
+        threshold.  Only relevant if the frequency map is provided at setup.
+    """
+
+    threshold = config.Property(proptype=float, default=5.0)
+    max_frac_freq = config.Property(proptype=float, default=0.1)
+
+    def setup(self, freq_map=None):
+        """Determine the frequencies handled by each FPGA motherboard slot.
+
+        If a motherboard decorrelates, then all frequencies transmitted by
+        that slot will be affected.
+
+        Parameters
+        ----------
+        freq_map : FrequencyMap
+            The mapping between frequency bin and [crate, slot, link]
+            as a function of time.
+        """
+
+        self._set_slot_freqs(freq_map)
+
+    def process(self, data, inputmap):
+        """Create the mask.
+
+        Parameters
+        ----------
+        data : TimeStream
+            Visibilites before averaging over cylinders.
+        inputmap : list of :class:`CorrInput`
+            A list describing the inputs in data.
+
+        Returns
+        -------
+        out : RFIMask
+            Mask with True indicating that a cylinder decorrelated
+            at that frequency and time.
+        """
+
+        from draco.analysis.ringmapmaker import find_grid_indices
+
+        # Distribute over frequencies
+        data.redistribute("freq")
+
+        # Get polarisations of feeds
+        pol = tools.get_feed_polarisations(inputmap)
+
+        # Get positions of feeds
+        pos = tools.get_feed_positions(inputmap)
+
+        # Get cylinder each feed is on
+        cyl = np.array(
+            [chr(63 + inp.cyl) if tools.is_chime(inp) else "0" for inp in inputmap]
+        )
+        ucyl = np.unique(cyl[cyl != "0"])
+        ncyl = ucyl.size
+
+        # Make sure that none of our typical products reference non-array feeds
+        stack_new, stack_flag = tools.redefine_stack_index_map(
+            inputmap, data.prod, data.stack, data.reverse_map["stack"]
+        )
+
+        valid_stack = np.flatnonzero(stack_flag)
+        ninvalid = stack_new.size - valid_stack.size
+
+        if ninvalid > 0:
+            stack_new = stack_new[valid_stack]
+            self.log.info(
+                "Could not find appropriate reference inputs for "
+                f"{ninvalid:0.0f} stacked products.  Ignoring these "
+                "products in decorrelated cylinder calculation."
+            )
+
+        t = data.prod[stack_new["prod"]]
+
+        prodstack = t.copy()
+        conj = stack_new["conjugate"]
+        prodstack["input_a"] = np.where(conj, t["input_b"], t["input_a"])
+        prodstack["input_b"] = np.where(conj, t["input_a"], t["input_b"])
+
+        # Calculate baseline distance and polarisation pair
+        index_a = prodstack["input_a"]
+        index_b = prodstack["input_b"]
```

Comment:
I guess you can't use `data.prodstack` in here because pre-collation it occasionally gives you a pair of feeds where one doesn't actually map to a real CHIME input?

Reply:
Yeah exactly.  I have to first use `tools.redefine_stack_index_map` to generate a stack index map that only references CHIME inputs and then copy over the code from `prodstack` get the products and perform the conjugation.

Reply:
Yeah I think that would be nice.  Are you envisioning a separate pipeline task or doing this in `andata` / `LoadCorrDataFiles`?

It does require an input map queried from the layout database.  Perhaps an alternative that does not require the layout database would be to use products whose inputs are not always flagged based on the `input_flags`.